File: swift/Sources/ProjectNameSDK/core/BluefinTecsUserBackofficeSDK.swift
// BluefinTecsUserBackoffice SDK client.

import Foundation

public final class BluefinTecsUserBackofficeSDK {
  public var mode = "live"
  private var options: VMap = VMap()
  private let utility: Utility
  public var features: [BaseFeature] = []
  private var rootctx: Context!

  public init(_ optionsIn: VMap? = nil) {
    utility = Utility()

    let config = SdkConfig.makeConfig()

    var ctxmap: [String: Any?] = [
      "client": self,
      "utility": utility,
      "config": config,
      "shared": VMap(),
    ]
    if let o = optionsIn { ctxmap["options"] = o }

    rootctx = utility.makeContext(ctxmap, nil)

    options = utility.makeOptions(rootctx)

    if gpath(options, "feature", "test", "active") == .bool(true) {
      mode = "test"
    }

    rootctx.options = options

    // Add features in the resolved order (makeOptions puts an explicit list
    // order first, else defaults to test-first). Ordering matters: the `test`
    // feature installs the base mock transport and the transport features
    // (retry/cache/netsim/proxy/ratelimit) wrap whatever is current, so `test`
    // must be added before them to sit at the base of the chain.
    let featureOpts = gp(options, "feature").asMap ?? VMap()
    if let featureOrder = gpath(options, "__derived__", "featureorder").asList {
      for fnameVal in featureOrder.items {
        let fname = fnameVal.asString ?? ""
        if fname != "", let fopts = gp(featureOpts, fname).asMap,
          fopts.entries["active"]?.asBool == true {
          utility.featureAdd(rootctx, SdkConfig.makeFeature(fname))
        }
      }
    }

    // Add extension features.
    if let extList = gp(options, "extend").asList {
      for f in extList.items {
        if let feat = f.asNative as? BaseFeature {
          utility.featureAdd(rootctx, feat)
        }
      }
    }

    // Initialize features.
    for f in features {
      utility.featureInit(rootctx, f)
    }

    utility.featureHook(rootctx, "PostConstruct")
  }

  public func optionsMap() -> VMap {
    return clone(.map(options)).asMap ?? VMap()
  }

  public func getUtility() -> Utility {
    return Utility.copy(utility)
  }

  public func getRootCtx() -> Context {
    return rootctx
  }

  public func prepare(_ fetchargsIn: VMap?) throws -> VMap {
    let utility = self.utility

    let fetchargs = fetchargsIn ?? VMap()

    let ctrl = gp(fetchargs, "ctrl").asMap ?? VMap()

    let ctx = utility.makeContext(["opname": "prepare", "ctrl": ctrl], rootctx)

    let options = self.options

    let path = gp(fetchargs, "path").asString ?? ""
    var method = gp(fetchargs, "method").asString ?? ""
    if method == "" { method = "GET" }

    let pathParams = gp(fetchargs, "params").asMap ?? VMap()
    let query = gp(fetchargs, "query").asMap ?? VMap()

    let headers = utility.prepareHeaders(ctx)

    let basev = gp(options, "base").asString ?? ""
    let prefix = gp(options, "prefix").asString ?? ""
    let suffix = gp(options, "suffix").asString ?? ""

    let specmap = VMap()
    specmap.entries["base"] = .string(basev)
    specmap.entries["prefix"] = .string(prefix)
    specmap.entries["suffix"] = .string(suffix)
    specmap.entries["path"] = .string(path)
    specmap.entries["method"] = .string(method)
    specmap.entries["params"] = .map(pathParams)
    specmap.entries["query"] = .map(query)
    specmap.entries["headers"] = .map(headers)
    specmap.entries["body"] = gp(fetchargs, "body")
    specmap.entries["step"] = .string("start")
    ctx.spec = Spec(specmap)

    // Merge user-provided headers.
    if let uhm = gp(fetchargs, "headers").asMap {
      for (k, v) in uhm.entries {
        ctx.spec!.headers.entries[k] = v
      }
    }

    _ = try utility.prepareAuth(ctx)

    return try utility.makeFetchDef(ctx)
  }

  public func direct(_ fetchargsIn: VMap?) -> VMap {
    let utility = self.utility

    let fetchdef: VMap
    do {
      fetchdef = try prepare(fetchargsIn)
    } catch {
      let r = VMap()
      r.entries["ok"] = .bool(false)
      r.entries["err"] = .nat(error)
      return r
    }

    let fetchargs = fetchargsIn ?? VMap()
    let ctrl = gp(fetchargs, "ctrl").asMap ?? VMap()

    let ctx = utility.makeContext(["opname": "direct", "ctrl": ctrl], rootctx)

    let url = gp(fetchdef, "url").asString ?? ""

    let fetched: Value
    do {
      fetched = try utility.fetcher(ctx, url, fetchdef)
    } catch {
      let r = VMap()
      r.entries["ok"] = .bool(false)
      r.entries["err"] = .nat(error)
      return r
    }

    if isNil(fetched) {
      let r = VMap()
      r.entries["ok"] = .bool(false)
      r.entries["err"] = .nat(ctx.makeError("direct_no_response", "response: undefined"))
      return r
    }

    if let fm = fetched.asMap {
      let status = toInt(gp(fm, "status"))
      let headers = gp(fm, "headers")

      // No-body responses (204, 304) and explicit zero content-length must
      // skip JSON parsing.
      var contentLength = ""
      if let hm = headers.asMap, let cl = hm.entries["content-length"], !isNil(cl) {
        contentLength = stringify(cl)
      }
      let noBody = status == 204 || status == 304 || contentLength == "0"

      var jsonData: Value = .noval
      if !noBody, let jf = gp(fm, "json").asNative as? NativeCall0 {
        jsonData = jf()
      }

      let r = VMap()
      r.entries["ok"] = .bool(status >= 200 && status < 300)
      r.entries["status"] = .int(Int64(status))
      r.entries["headers"] = headers
      r.entries["data"] = jsonData
      return r
    }

    let r = VMap()
    r.entries["ok"] = .bool(false)
    r.entries["err"] = .nat(ctx.makeError("direct_invalid", "invalid response type"))
    return r
  }


  // OutputActivateDigitalModule returns a OutputActivateDigitalModule entity bound to this client.
  // Idiomatic usage: try client.OutputActivateDigitalModule().list(nil) or
  // try client.OutputActivateDigitalModule().load(vm(("id", .string("..."))), nil).
  public func OutputActivateDigitalModule(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return OutputActivateDigitalModuleEntity(self, entopts)
  }

  // OutputActivatePortalModule returns a OutputActivatePortalModule entity bound to this client.
  // Idiomatic usage: try client.OutputActivatePortalModule().list(nil) or
  // try client.OutputActivatePortalModule().load(vm(("id", .string("..."))), nil).
  public func OutputActivatePortalModule(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return OutputActivatePortalModuleEntity(self, entopts)
  }

  // OutputActivateStoreModule returns a OutputActivateStoreModule entity bound to this client.
  // Idiomatic usage: try client.OutputActivateStoreModule().list(nil) or
  // try client.OutputActivateStoreModule().load(vm(("id", .string("..."))), nil).
  public func OutputActivateStoreModule(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return OutputActivateStoreModuleEntity(self, entopts)
  }

  // OutputActivateUser returns a OutputActivateUser entity bound to this client.
  // Idiomatic usage: try client.OutputActivateUser().list(nil) or
  // try client.OutputActivateUser().load(vm(("id", .string("..."))), nil).
  public func OutputActivateUser(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return OutputActivateUserEntity(self, entopts)
  }

  // OutputAssignRole returns a OutputAssignRole entity bound to this client.
  // Idiomatic usage: try client.OutputAssignRole().list(nil) or
  // try client.OutputAssignRole().load(vm(("id", .string("..."))), nil).
  public func OutputAssignRole(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return OutputAssignRoleEntity(self, entopts)
  }

  // OutputChangeLogo returns a OutputChangeLogo entity bound to this client.
  // Idiomatic usage: try client.OutputChangeLogo().list(nil) or
  // try client.OutputChangeLogo().load(vm(("id", .string("..."))), nil).
  public func OutputChangeLogo(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return OutputChangeLogoEntity(self, entopts)
  }

  // OutputCreateMandator returns a OutputCreateMandator entity bound to this client.
  // Idiomatic usage: try client.OutputCreateMandator().list(nil) or
  // try client.OutputCreateMandator().load(vm(("id", .string("..."))), nil).
  public func OutputCreateMandator(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return OutputCreateMandatorEntity(self, entopts)
  }

  // OutputCreateServiceUser returns a OutputCreateServiceUser entity bound to this client.
  // Idiomatic usage: try client.OutputCreateServiceUser().list(nil) or
  // try client.OutputCreateServiceUser().load(vm(("id", .string("..."))), nil).
  public func OutputCreateServiceUser(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return OutputCreateServiceUserEntity(self, entopts)
  }

  // OutputDeactivateUser returns a OutputDeactivateUser entity bound to this client.
  // Idiomatic usage: try client.OutputDeactivateUser().list(nil) or
  // try client.OutputDeactivateUser().load(vm(("id", .string("..."))), nil).
  public func OutputDeactivateUser(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return OutputDeactivateUserEntity(self, entopts)
  }

  // OutputGetKycDocument returns a OutputGetKycDocument entity bound to this client.
  // Idiomatic usage: try client.OutputGetKycDocument().list(nil) or
  // try client.OutputGetKycDocument().load(vm(("id", .string("..."))), nil).
  public func OutputGetKycDocument(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return OutputGetKycDocumentEntity(self, entopts)
  }

  // OutputGetLogo returns a OutputGetLogo entity bound to this client.
  // Idiomatic usage: try client.OutputGetLogo().list(nil) or
  // try client.OutputGetLogo().load(vm(("id", .string("..."))), nil).
  public func OutputGetLogo(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return OutputGetLogoEntity(self, entopts)
  }

  // OutputListOfAvailableRole returns a OutputListOfAvailableRole entity bound to this client.
  // Idiomatic usage: try client.OutputListOfAvailableRole().list(nil) or
  // try client.OutputListOfAvailableRole().load(vm(("id", .string("..."))), nil).
  public func OutputListOfAvailableRole(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return OutputListOfAvailableRoleEntity(self, entopts)
  }

  // OutputListOfMandator returns a OutputListOfMandator entity bound to this client.
  // Idiomatic usage: try client.OutputListOfMandator().list(nil) or
  // try client.OutputListOfMandator().load(vm(("id", .string("..."))), nil).
  public func OutputListOfMandator(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return OutputListOfMandatorEntity(self, entopts)
  }

  // OutputListOfModule returns a OutputListOfModule entity bound to this client.
  // Idiomatic usage: try client.OutputListOfModule().list(nil) or
  // try client.OutputListOfModule().load(vm(("id", .string("..."))), nil).
  public func OutputListOfModule(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return OutputListOfModuleEntity(self, entopts)
  }

  // OutputListOfRoleGroup returns a OutputListOfRoleGroup entity bound to this client.
  // Idiomatic usage: try client.OutputListOfRoleGroup().list(nil) or
  // try client.OutputListOfRoleGroup().load(vm(("id", .string("..."))), nil).
  public func OutputListOfRoleGroup(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return OutputListOfRoleGroupEntity(self, entopts)
  }

  // OutputListOfTransactionsHistory returns a OutputListOfTransactionsHistory entity bound to this client.
  // Idiomatic usage: try client.OutputListOfTransactionsHistory().list(nil) or
  // try client.OutputListOfTransactionsHistory().load(vm(("id", .string("..."))), nil).
  public func OutputListOfTransactionsHistory(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return OutputListOfTransactionsHistoryEntity(self, entopts)
  }

  // OutputListOfUser returns a OutputListOfUser entity bound to this client.
  // Idiomatic usage: try client.OutputListOfUser().list(nil) or
  // try client.OutputListOfUser().load(vm(("id", .string("..."))), nil).
  public func OutputListOfUser(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return OutputListOfUserEntity(self, entopts)
  }

  // OutputProvideCredential returns a OutputProvideCredential entity bound to this client.
  // Idiomatic usage: try client.OutputProvideCredential().list(nil) or
  // try client.OutputProvideCredential().load(vm(("id", .string("..."))), nil).
  public func OutputProvideCredential(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return OutputProvideCredentialEntity(self, entopts)
  }

  // OutputRegisterUser returns a OutputRegisterUser entity bound to this client.
  // Idiomatic usage: try client.OutputRegisterUser().list(nil) or
  // try client.OutputRegisterUser().load(vm(("id", .string("..."))), nil).
  public func OutputRegisterUser(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return OutputRegisterUserEntity(self, entopts)
  }

  // OutputRemoveRole returns a OutputRemoveRole entity bound to this client.
  // Idiomatic usage: try client.OutputRemoveRole().list(nil) or
  // try client.OutputRemoveRole().load(vm(("id", .string("..."))), nil).
  public func OutputRemoveRole(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return OutputRemoveRoleEntity(self, entopts)
  }

  // OutputResendLink returns a OutputResendLink entity bound to this client.
  // Idiomatic usage: try client.OutputResendLink().list(nil) or
  // try client.OutputResendLink().load(vm(("id", .string("..."))), nil).
  public func OutputResendLink(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return OutputResendLinkEntity(self, entopts)
  }

  // OutputResetPassword returns a OutputResetPassword entity bound to this client.
  // Idiomatic usage: try client.OutputResetPassword().list(nil) or
  // try client.OutputResetPassword().load(vm(("id", .string("..."))), nil).
  public func OutputResetPassword(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return OutputResetPasswordEntity(self, entopts)
  }

  // OutputUpdateConsumer returns a OutputUpdateConsumer entity bound to this client.
  // Idiomatic usage: try client.OutputUpdateConsumer().list(nil) or
  // try client.OutputUpdateConsumer().load(vm(("id", .string("..."))), nil).
  public func OutputUpdateConsumer(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return OutputUpdateConsumerEntity(self, entopts)
  }

  // OutputUpdateProfile returns a OutputUpdateProfile entity bound to this client.
  // Idiomatic usage: try client.OutputUpdateProfile().list(nil) or
  // try client.OutputUpdateProfile().load(vm(("id", .string("..."))), nil).
  public func OutputUpdateProfile(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return OutputUpdateProfileEntity(self, entopts)
  }

  // Version returns a Version entity bound to this client.
  // Idiomatic usage: try client.Version().list(nil) or
  // try client.Version().load(vm(("id", .string("..."))), nil).
  public func Version(_ entopts: VMap? = nil) -> BluefinTecsUserBackofficeEntityBase {
    return VersionEntity(self, entopts)
  }


  public static func testSDK(_ testoptsIn: VMap?, _ sdkoptsIn: VMap?) -> BluefinTecsUserBackofficeSDK {
    let sdkopts = clone(.map(sdkoptsIn ?? VMap())).asMap ?? VMap()

    let testopts = clone(.map(testoptsIn ?? VMap())).asMap ?? VMap()
    testopts.entries["active"] = .bool(true)

    _ = setpath(.map(sdkopts), jtp("feature", "test"), .map(testopts))

    let sdk = BluefinTecsUserBackofficeSDK(sdkopts)
    sdk.mode = "test"
    return sdk
  }
}

File: js/src/BluefinTecsUserBackofficeSDK.js
// BluefinTecsUserBackoffice Js SDK

const { OutputActivateDigitalModuleEntity } = require('./entity/OutputActivateDigitalModuleEntity')
const { OutputActivatePortalModuleEntity } = require('./entity/OutputActivatePortalModuleEntity')
const { OutputActivateStoreModuleEntity } = require('./entity/OutputActivateStoreModuleEntity')
const { OutputActivateUserEntity } = require('./entity/OutputActivateUserEntity')
const { OutputAssignRoleEntity } = require('./entity/OutputAssignRoleEntity')
const { OutputChangeLogoEntity } = require('./entity/OutputChangeLogoEntity')
const { OutputCreateMandatorEntity } = require('./entity/OutputCreateMandatorEntity')
const { OutputCreateServiceUserEntity } = require('./entity/OutputCreateServiceUserEntity')
const { OutputDeactivateUserEntity } = require('./entity/OutputDeactivateUserEntity')
const { OutputGetKycDocumentEntity } = require('./entity/OutputGetKycDocumentEntity')
const { OutputGetLogoEntity } = require('./entity/OutputGetLogoEntity')
const { OutputListOfAvailableRoleEntity } = require('./entity/OutputListOfAvailableRoleEntity')
const { OutputListOfMandatorEntity } = require('./entity/OutputListOfMandatorEntity')
const { OutputListOfModuleEntity } = require('./entity/OutputListOfModuleEntity')
const { OutputListOfRoleGroupEntity } = require('./entity/OutputListOfRoleGroupEntity')
const { OutputListOfTransactionsHistoryEntity } = require('./entity/OutputListOfTransactionsHistoryEntity')
const { OutputListOfUserEntity } = require('./entity/OutputListOfUserEntity')
const { OutputProvideCredentialEntity } = require('./entity/OutputProvideCredentialEntity')
const { OutputRegisterUserEntity } = require('./entity/OutputRegisterUserEntity')
const { OutputRemoveRoleEntity } = require('./entity/OutputRemoveRoleEntity')
const { OutputResendLinkEntity } = require('./entity/OutputResendLinkEntity')
const { OutputResetPasswordEntity } = require('./entity/OutputResetPasswordEntity')
const { OutputUpdateConsumerEntity } = require('./entity/OutputUpdateConsumerEntity')
const { OutputUpdateProfileEntity } = require('./entity/OutputUpdateProfileEntity')
const { VersionEntity } = require('./entity/VersionEntity')


const { inspect } = require('node:util')

const { config } = require('./Config')
const { Utility } = require('./utility/Utility')
const { BluefinTecsUserBackofficeEntityBase } = require('./BluefinTecsUserBackofficeEntityBase')


const { BaseFeature } = require('./feature/base/BaseFeature')


const stdutil = new Utility()


class BluefinTecsUserBackofficeSDK {
  _mode = 'live'
  _options
  _utility = new Utility()
  _features
  _rootctx

  constructor(options) {

    this._rootctx = this._utility.makeContext({
      client: this,
      utility: this._utility,
      config,
      options,
      shared: new WeakMap()
    })

    this._options = this._utility.makeOptions(this._rootctx)

    const struct = this._utility.struct
    const getpath = struct.getpath

    if (true === getpath(this._options.feature, 'test.active')) {
      this._mode = 'test'
    }

    this._rootctx.options = this._options

    this._features = []

    const featureAdd = this._utility.featureAdd
    const featureInit = this._utility.featureInit

    // Add features in the resolved order (makeOptions puts an explicit
    // array order first, else defaults to test-first). Ordering matters:
    // the `test` feature installs the base mock transport and the transport
    // features (retry/cache/netsim/proxy/ratelimit) wrap whatever is current,
    // so `test` must be added before them to sit at the base of the chain.
    const extend = this._options.extend || []

    const featureorder = getpath(this._options, '__derived__.featureorder') || []
    for (const fname of featureorder) {
      const fopts = this._options.feature[fname] || {}
      if (fopts.active) {
        // An active name with no generated class is legal when an
        // extend-supplied instance carries that name (station's adopt
        // path): the instance is added below, positioned by its own
        // __after__ entry, so skip it here rather than fail construction.
        if (!this._rootctx.config.hasFeature(fname) &&
          extend.some((f) => fname === f.name)) {
          continue
        }
        featureAdd(this._rootctx, this._rootctx.config.makeFeature(fname))
      }
    }

    for (let f of extend) {
      featureAdd(this._rootctx, f)
    }

    for (let f of this._features) {
      featureInit(this._rootctx, f)
    }

    const featureHook = this._utility.featureHook
    featureHook(this._rootctx, 'PostConstruct')
  }


  options() {
    return this._utility.struct.clone(this._options)
  }


  utility() {
    return this._utility.struct.clone(this._utility)
  }


  async prepare(fetchargs) {
    const utility = this._utility
    const struct = utility.struct
    const clone = struct.clone

    const {
      makeContext,
      makeFetchDef,
      prepareHeaders,
      prepareAuth,
    } = utility

    fetchargs = fetchargs || {}

    let ctx = makeContext({
      opname: 'prepare',
      ctrl: fetchargs.ctrl || {},
    }, this._rootctx)

    const options = this._options

    // Build spec directly from SDK options + user-provided fetch args.
    const spec = {
      base: options.base,
      prefix: options.prefix,
      suffix: options.suffix,
      path: fetchargs.path || '',
      method: fetchargs.method || 'GET',
      params: fetchargs.params || {},
      query: fetchargs.query || {},
      headers: prepareHeaders(ctx),
      body: fetchargs.body,
      step: 'start',
    }

    ctx.spec = spec

    // Merge user-provided headers over SDK defaults.
    if (fetchargs.headers) {
      const uheaders = fetchargs.headers
      for (let key in uheaders) {
        spec.headers[key] = uheaders[key]
      }
    }

    // Apply SDK auth (apikey, auth prefix, etc.)
    const authResult = prepareAuth(ctx)
    if (authResult instanceof Error) {
      return authResult
    }

    return makeFetchDef(ctx)
  }


  // Raw endpoint access is operator-controllable, like every entity op.
  // Blocking it means denying BOTH the 'direct' and 'graphql' tokens, since
  // either one reaches the same endpoint.
  async direct(fetchargs) {
    if (!this._options.allow.op.includes('direct')) {
      return {
        ok: false,
        err: new Error('BluefinTecsUserBackofficeSDK: direct: operation not allowed by' +
          ' SDK option allow.op value: "' + this._options.allow.op + '"'),
      }
    }

    return this._rawRequest(fetchargs)
  }


  // Ungated request path shared by direct() and graphql(), each of which
  // checks its own allow.op token first. Private, rather than a flag on
  // fetchargs: a caller-supplied marker would let anyone opt straight back
  // out of the gate by passing it.
  async _rawRequest(fetchargs) {
    const utility = this._utility

    const fetcher = utility.fetcher
    const makeContext = utility.makeContext

    const fetchdef = await this.prepare(fetchargs)
    if (fetchdef instanceof Error) {
      return fetchdef
    }

    let ctx = makeContext({
      opname: 'direct',
      ctrl: (fetchargs || {}).ctrl || {},
    }, this._rootctx)

    try {
      const fetched = await fetcher(ctx, fetchdef.url, fetchdef)

      if (null == fetched) {
        return { ok: false, err: ctx.error('direct_no_response', 'response: undefined') }
      }
      else if (fetched instanceof Error) {
        return { ok: false, err: fetched }
      }

      const status = fetched.status

      // No body responses (204 No Content, 304 Not Modified) and explicit
      // zero content-length must skip JSON parsing — fetched.json() would
      // throw `Unexpected end of JSON input` on an empty body.
      const headers = fetched.headers
      const contentLength = headers && 'function' === typeof headers.get
        ? headers.get('content-length')
        : (headers || {})['content-length']
      const noBody = 204 === status || 304 === status || '0' === String(contentLength)

      let json = undefined
      if (!noBody) {
        try {
          json = 'function' === typeof fetched.json ? await fetched.json() : fetched.json
        }
        catch (parseErr) {
          // Body wasn't valid JSON — surface the raw response rather than
          // throwing. data stays undefined; callers can inspect status/headers.
          json = undefined
        }
      }

      return {
        ok: status >= 200 && status < 300,
        status,
        headers: fetched.headers,
        data: json,
      }
    }
    catch (err) {
      return { ok: false, err }
    }
  }



  // Raw GraphQL access: the pressure valve that makes the generated
  // surface's deliberate omissions (per-call selection sets, typed filter
  // builders, batching, subscriptions) livable — the whole schema stays
  // reachable.
  //
  // Thin wrapper over the same prepare/fetch path `direct` uses, with the
  // one thing raw `direct` cannot do for GraphQL: a GraphQL failure rides
  // HTTP 200 as a top-level `errors` array, so status alone would report a
  // failed query as ok.
  //
  // NOTE: like `direct`, this bypasses the feature pipeline — no retry,
  // ratelimit or paging features apply.
  async graphql(query, variables, ctrl) {
    const options = this._options

    if (!options.allow.op.includes('graphql')) {
      return {
        ok: false,
        err: new Error('BluefinTecsUserBackofficeSDK: graphql: operation not allowed by' +
          ' SDK option allow.op value: "' + options.allow.op + '"'),
      }
    }

    const res = await this._rawRequest({
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: { query, variables: variables || {} },
      ctrl,
    })

    if (res instanceof Error) {
      return res
    }

    // Errors are read BEFORE any status check: a GraphQL parse or validation
    // failure comes back as HTTP 400 carrying the standard { errors: [...] }
    // body, and the raw path represents a non-2xx as { ok: false } with no
    // err — so returning early on status would discard the server's own
    // diagnostics, which are the only useful part of that response.
    const errors = null == res.data ? undefined : res.data.errors

    if (null != errors && Array.isArray(errors) && 0 < errors.length) {
      const first = errors[0] || {}
      const err = new Error('BluefinTecsUserBackofficeSDK: graphql: ' +
        (first.message || 'graphql error'))
      err.graphql = errors
      return { ok: false, status: res.status, headers: res.headers, err, data: res.data }
    }

    return res
  }



  // Entity access: `client.OutputActivateDigitalModule().list()` / `client.OutputActivateDigitalModule().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  OutputActivateDigitalModule(entopts) {
    const self = this
    return new OutputActivateDigitalModuleEntity(self, entopts)
  }


  // Entity access: `client.OutputActivatePortalModule().list()` / `client.OutputActivatePortalModule().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  OutputActivatePortalModule(entopts) {
    const self = this
    return new OutputActivatePortalModuleEntity(self, entopts)
  }


  // Entity access: `client.OutputActivateStoreModule().list()` / `client.OutputActivateStoreModule().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  OutputActivateStoreModule(entopts) {
    const self = this
    return new OutputActivateStoreModuleEntity(self, entopts)
  }


  // Entity access: `client.OutputActivateUser().list()` / `client.OutputActivateUser().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  OutputActivateUser(entopts) {
    const self = this
    return new OutputActivateUserEntity(self, entopts)
  }


  // Entity access: `client.OutputAssignRole().list()` / `client.OutputAssignRole().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  OutputAssignRole(entopts) {
    const self = this
    return new OutputAssignRoleEntity(self, entopts)
  }


  // Entity access: `client.OutputChangeLogo().list()` / `client.OutputChangeLogo().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  OutputChangeLogo(entopts) {
    const self = this
    return new OutputChangeLogoEntity(self, entopts)
  }


  // Entity access: `client.OutputCreateMandator().list()` / `client.OutputCreateMandator().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  OutputCreateMandator(entopts) {
    const self = this
    return new OutputCreateMandatorEntity(self, entopts)
  }


  // Entity access: `client.OutputCreateServiceUser().list()` / `client.OutputCreateServiceUser().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  OutputCreateServiceUser(entopts) {
    const self = this
    return new OutputCreateServiceUserEntity(self, entopts)
  }


  // Entity access: `client.OutputDeactivateUser().list()` / `client.OutputDeactivateUser().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  OutputDeactivateUser(entopts) {
    const self = this
    return new OutputDeactivateUserEntity(self, entopts)
  }


  // Entity access: `client.OutputGetKycDocument().list()` / `client.OutputGetKycDocument().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  OutputGetKycDocument(entopts) {
    const self = this
    return new OutputGetKycDocumentEntity(self, entopts)
  }


  // Entity access: `client.OutputGetLogo().list()` / `client.OutputGetLogo().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  OutputGetLogo(entopts) {
    const self = this
    return new OutputGetLogoEntity(self, entopts)
  }


  // Entity access: `client.OutputListOfAvailableRole().list()` / `client.OutputListOfAvailableRole().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  OutputListOfAvailableRole(entopts) {
    const self = this
    return new OutputListOfAvailableRoleEntity(self, entopts)
  }


  // Entity access: `client.OutputListOfMandator().list()` / `client.OutputListOfMandator().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  OutputListOfMandator(entopts) {
    const self = this
    return new OutputListOfMandatorEntity(self, entopts)
  }


  // Entity access: `client.OutputListOfModule().list()` / `client.OutputListOfModule().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  OutputListOfModule(entopts) {
    const self = this
    return new OutputListOfModuleEntity(self, entopts)
  }


  // Entity access: `client.OutputListOfRoleGroup().list()` / `client.OutputListOfRoleGroup().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  OutputListOfRoleGroup(entopts) {
    const self = this
    return new OutputListOfRoleGroupEntity(self, entopts)
  }


  // Entity access: `client.OutputListOfTransactionsHistory().list()` / `client.OutputListOfTransactionsHistory().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  OutputListOfTransactionsHistory(entopts) {
    const self = this
    return new OutputListOfTransactionsHistoryEntity(self, entopts)
  }


  // Entity access: `client.OutputListOfUser().list()` / `client.OutputListOfUser().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  OutputListOfUser(entopts) {
    const self = this
    return new OutputListOfUserEntity(self, entopts)
  }


  // Entity access: `client.OutputProvideCredential().list()` / `client.OutputProvideCredential().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  OutputProvideCredential(entopts) {
    const self = this
    return new OutputProvideCredentialEntity(self, entopts)
  }


  // Entity access: `client.OutputRegisterUser().list()` / `client.OutputRegisterUser().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  OutputRegisterUser(entopts) {
    const self = this
    return new OutputRegisterUserEntity(self, entopts)
  }


  // Entity access: `client.OutputRemoveRole().list()` / `client.OutputRemoveRole().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  OutputRemoveRole(entopts) {
    const self = this
    return new OutputRemoveRoleEntity(self, entopts)
  }


  // Entity access: `client.OutputResendLink().list()` / `client.OutputResendLink().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  OutputResendLink(entopts) {
    const self = this
    return new OutputResendLinkEntity(self, entopts)
  }


  // Entity access: `client.OutputResetPassword().list()` / `client.OutputResetPassword().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  OutputResetPassword(entopts) {
    const self = this
    return new OutputResetPasswordEntity(self, entopts)
  }


  // Entity access: `client.OutputUpdateConsumer().list()` / `client.OutputUpdateConsumer().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  OutputUpdateConsumer(entopts) {
    const self = this
    return new OutputUpdateConsumerEntity(self, entopts)
  }


  // Entity access: `client.OutputUpdateProfile().list()` / `client.OutputUpdateProfile().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  OutputUpdateProfile(entopts) {
    const self = this
    return new OutputUpdateProfileEntity(self, entopts)
  }


  // Entity access: `client.Version().list()` / `client.Version().load({ id })`.
  // The argument is the entity OPTIONS object (passed to the entity
  // constructor as entopts), not initial entity data.
  Version(entopts) {
    const self = this
    return new VersionEntity(self, entopts)
  }




  static test(testoptsarg, sdkoptsarg) {
    const struct = stdutil.struct
    const setpath = struct.setpath
    const getdef = struct.getdef
    const clone = struct.clone
    const setprop = struct.setprop

    const sdkopts = getdef(clone(sdkoptsarg), {})
    const testopts = getdef(clone(testoptsarg), {})
    setprop(testopts, 'active', true)
    setpath(sdkopts, 'feature.test', testopts)

    const testsdk = new BluefinTecsUserBackofficeSDK(sdkopts)
    testsdk._mode = 'test'

    return testsdk
  }


  tester(testopts, sdkopts) {
    return BluefinTecsUserBackofficeSDK.test(testopts, sdkopts)
  }


  toJSON() {
    return { name: 'BluefinTecsUserBackoffice' }
  }

  toString() {
    return 'BluefinTecsUserBackoffice ' + this._utility.struct.jsonify(this.toJSON())
  }

  [inspect.custom]() {
    return this.toString()
  }

}




const SDK = BluefinTecsUserBackofficeSDK


module.exports = {
  stdutil,
  config,

  BaseFeature,
  BluefinTecsUserBackofficeEntityBase,

  BluefinTecsUserBackofficeSDK,
  SDK,
}

